### PR TITLE
Layers

### PR DESCRIPTION
**Pull request by krenzland**
Adds the possibility of creating layouts and views with different layers.
Useful e.g. for the implementation of ghost/interior layers.
I made heavy use of the builder pattern; it is type safe insofar as it only compiles once all attributes are set. Results in template hell but it's worth it.
I've added tests for all new features.

I've also removed a clang-tidy warning, as using it leads to a contradiction with gcc warnings.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -7,7 +7,7 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
 
 enable_testing()
-include(CTest)
+enable_testing()
```
Comment (uphoffc) [inline CMakeLists.txt]: Twice?

File: include/mneme/plan.hpp [hunk 2]
```diff
@@ -2,24 +2,102 @@
...
     std::vector<std::size_t> dofs;
 };
 
+class Layer {
+public:
+    constexpr Layer() = default;
+    constexpr Layer(std::size_t numElements, std::size_t offset)
+        : numElements(numElements), offset(offset) {}
+    size_t numElements = 0;
+    size_t offset = 0;
```
Comment (uphoffc) [inline include/mneme/plan.hpp]: size_t vs std::size_t?

File: include/mneme/plan.hpp [hunk 3]
```diff
@@ -2,24 +2,102 @@
...
+        this->numElements = otherPlan.numElements;
+        this->plan = otherPlan.plan;
+        std::apply(
+            [&](auto&&... args) {
+                (((std::get<typename std::remove_reference_t<std::remove_const_t<decltype(args)>>>(
+                      layers)) =
+                      std::get<typename std::remove_reference_t<decltype(args)>>(otherPlan.layers)),
+                 ...);
+            },
+            otherPlan.layers);
```
Comment (uphoffc) [inline include/mneme/plan.hpp:82]: Could you comment what this thing does?

File: test/mneme.cpp [hunk 4]
```diff
@@ -78,3 +84,100 @@ TEST_CASE("Data structure works") {
...
+    constexpr size_t numInterior = 100;
+    constexpr size_t numCopy = 30;
+    constexpr size_t numGhost = 20;
+
+    constexpr size_t numMaterial = 1;
+    constexpr auto numMaterialFunc = [numMaterial](auto) { return numMaterial; };
+    const auto localPlan = LayeredPlan()
+                               .withDofs<Interior>(numInterior, numMaterialFunc)
+                               .withDofs<Copy>(numInterior, numMaterialFunc)
+                               .withDofs<Ghost>(numInterior, numMaterialFunc);
```
Comment (uphoffc) [inline test/mneme.cpp:124]: Da sind jetzt immer alle interior layer hintereinander, und dann alle copy usw?

Comment (uphoffc) [reply test/mneme.cpp:124]: Also das ist quasi für einen cluster, oder? Was macht man dann bei mehreren?

File: include/mneme/view.hpp [hunk 5]
```diff
@@ -91,6 +92,74 @@ template <typename Storage> class GeneralView {
...
+        return DenseView<std::remove_pointer_t<typename MaybeStorage_::type>>(
+            layout, *(maybeStorage.value), from, to);
+    }
+
+private:
+    MaybePlan maybePlan;
+    MaybeStorage maybeStorage;
+};
+
+constexpr auto createView() {
```
Comment (uphoffc) [inline include/mneme/view.hpp]: Vllt eher createViewFactory?

File: include/mneme/plan.hpp (hunk 2 above)
Comment (krenzland) [reply include/mneme/plan.hpp]: https://stackoverflow.com/questions/5813700/difference-between-size-t-and-stdsize-t
It's the same type but I agree - std::size_t is nicer

File: include/mneme/plan.hpp (hunk 3 above)
Comment (krenzland) [reply include/mneme/plan.hpp:82]: Magic.

It copies all entries from the old tuple to the new tuple. Not sure if I should add a comment here - it's a more or less basic abuse of std::apply plus a fold expression over the comma operator.
Maybe a comment would be a good idea...

File: test/mneme.cpp (hunk 4 above)
Comment (krenzland) [reply test/mneme.cpp:124]: Ja, genau. Spricht da was konkret dagegen? Bei unserem actor model haben wir ja auch einen actor jeweils für interior/copy/ghost.

Comment (uphoffc) [reply test/mneme.cpp:124]: Ne, denke nicht. Für das alte Layout vom Alex braucht man aber denke ich kontinuierlichen Speicher für alle

Comment (krenzland) [reply test/mneme.cpp:124]: Ja, du hast recht. Hab tatsächlich gar nicht an mehrere cluster gedacht. Ich bastel da mal ne Lösung.

File: include/mneme/plan.hpp [hunk 6]
```diff
@@ -2,24 +2,154 @@
...
+    std::size_t offset = 0;
+};
+
+class LayeredPlanBase {};
+
+template <typename... Layers> class LayeredPlan : public LayeredPlanBase {
+    template <typename... OtherLayers> friend class LayeredPlan;
+
+public:
+    using LayoutT = Displacements<std::size_t>;
```
Comment (uphoffc) [inline include/mneme/plan.hpp]: Make the name consistent? LayoutT vs LayoutType in line 19. (I'd prefer layout_t, but I won't argue about taste :))

File: include/mneme/plan.hpp [hunk 7]
```diff
@@ -2,24 +2,154 @@
...
+    LayeredPlan() : plan(0){};
```
Comment (uphoffc) [inline include/mneme/plan.hpp]: plan(0){}; -> plan(0) {}

File: include/mneme/plan.hpp [hunk 8]
```diff
@@ -2,24 +2,154 @@
...
+        for (std::size_t i = 0; i < newLayer.numElements; ++i) {
+            newPlan.plan.setDof(i + newLayer.offset, func(i));
+        }
+        return newPlan;
+    }
+
+    const LayoutT& getLayout() const {
+        // if (layout == std::nullopt) {
+        layout = plan.getLayout();
+        //}
```
Comment (uphoffc) [inline include/mneme/plan.hpp]: Why comments?

File: include/mneme/plan.hpp [hunk 9]
```diff
@@ -2,24 +2,154 @@
...
+    constexpr Layer() = default;
+    constexpr Layer(std::size_t numElements, std::size_t offset)
+        : numElements(numElements), offset(offset) {}
+    std::size_t numElements = 0;
+    std::size_t offset = 0;
+};
+
+class LayeredPlanBase {};
+
+template <typename... Layers> class LayeredPlan : public LayeredPlanBase {
```
Comment (uphoffc) [inline include/mneme/plan.hpp:57]: Could you add a short doxygen string which briefly describes what this class does and which includes a usage example?

**Review by uphoffc — APPROVED** [review]

**Pull request merged**